pyoozie/tags.py

```diff
@@ -494,6 +496,133 @@ def _xml(self, doc, tag, text):
+            on_error if on_error else on_next
+        )
+
+    @abc.abstractmethod
+    def _xml(self, doc, tag, text, on_next, on_error):
+        # type: (yattag.doc.Doc, yattag.doc.Doc.tag, yattag.doc.Doc.text, typing.Text, typing.Text) -> yattag.doc.Doc
+        raise NotImplementedError()
+
+    def __iter__(self):
+        # type: () -> typing.Generator[_WorkflowEntity, None, None]
+        yield self
+        if self.__on_error:
+            for action in self.__on_error:
+                yield action
+
+    def __bool__(self):  # type: () -> bool
+        return bool(list(self))
+
+    def __nonzero__(self):  # type: () -> bool
+        return self.__bool__()
+
+    def __repr__(self):  # type: () -> str
+        return str('{_class}({identifier})'.format(_class=type(self).__name__, identifier=self.identifier()))
+
+
+class Kill(_WorkflowEntity):
+    """Workflow graph terminal node(s) to end upon to indicate failure."""
+
+    def __init__(self, message, name=None):
+        # type: (typing.Text, typing.Optional[typing.Text]) -> None
+        super(Kill, self).__init__(xml_tag='kill', name=name)
+        self.message = message
+
+    def _xml(self, doc, tag, text, on_next, on_error):
+        # type: (yattag.doc.Doc, yattag.doc.Doc.tag, yattag.doc.Doc.text, typing.Text, typing.Text) -> yattag.doc.Doc
+        with tag(self.xml_tag, name=self.identifier()):
+            with tag('message'):
+                doc.text(self.message)
+        return doc
+
+
+ConcreteAction = typing.Union[Shell, SubWorkflow, Email]
+
+
+class Action(_WorkflowEntity):
+    """Workflow action nodes carrying concrete actions that perform an action."""
+
+    def __init__(
+            self,
+            action,               # type: ConcreteAction
+            name=None,            # type: typing.Optional[typing.Text]
+            credential=None,      # type: typing.Optional[typing.Text]
+            retry_max=None,       # type: typing.Optional[int]
+            retry_interval=None,  # type: typing.Optional[int]
+            on_error=None         # type: typing.Optional[_WorkflowEntity]
+    ):
+        # type: (...) -> None
+        super(Action, self).__init__(xml_tag='action', name=name, on_error=on_error)
+
+        # XML-document-related values
+        self.__action = action
+        self.__credential = credential
+        self.__retry_max = retry_max
+        self.__retry_interval = retry_interval
+
+    def credential(self):
+        # type: () -> typing.Optional[typing.Text]
+        return self.__credential
+
+    def _xml(self, doc, tag, text, on_next, on_error):
+        # type: (yattag.doc.Doc, yattag.doc.Doc.tag, yattag.doc.Doc.text, typing.Text, typing.Text) -> yattag.doc.Doc
+        _on_error = self._xml_and_get_on_error(doc, tag, text, on_next, on_error)
+
+        attributes = {
+            'name': self.identifier(),
+        }
+        if self.__credential:
+            attributes['cred'] = self.__credential
+        if self.__retry_max:
```

Comment:
Tested:
- If either or both are unspecified, Oozie will use [default values](url) (defined as `oozie.action.retries.max` or `oozie.action.retry.interval` in 4.2.0)
- If negative values are specified, they are accepted (unsure of how they're interpreted though but a workflow can complete successfully).